ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+        RequestInfo request = new RequestInfo(hostname, port, getRequest);
+        int correlationId = getRequest.getCorrelationId();
+        correlationIdToGetRequestInfo.put(correlationId, new GetRequestInfo(replicaId, time.milliseconds()));
+        correlationIdToGetChunk.put(correlationId, this);
+        requestRegistrationCallback.registerRequestToSend(GetBlobOperation.this, request);
+        state = ChunkState.InProgress;
+      }
+    }
+
+    /**
+     * Check if the operation on the chunk is eligible for completion, if so complete it.
+     */
+    void checkAndMaybeComplete() {
+      if (chunkOperationTracker.isDone()) {
+        if (chunkOperationTracker.hasSucceeded()) {
+          // override any previously set exceptions.
+          chunkException = null;
+        }
+        chunkCompleted = true;
+      }
+      if (chunkCompleted) {
+        operationException.set(chunkException);
+        state = ChunkState.Complete;
+      }
+    }
+
+    /**
+     * Handle the body of the response: Deserialize and add to the list of chunk buffers.
+     * @param payload the body of the response.
+     * @throws IOException if there is an IOException while deserializing the body.
+     * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
+     */
+    void handleBody(InputStream payload)
+        throws IOException, MessageFormatException {
+      if (!successfullyDeserialized) {
+        BlobData blobData = MessageFormatRecord.deserializeBlob(payload);
+        chunkIndexToBuffer.put(chunkIndex,
+            ByteBuffer.wrap(Utils.readBytesFromStream(blobData.getStream(), (int) blobData.getSize())));
+        numChunksRetrieved++;
+        successfullyDeserialized = true;
+      } else {
+        // If successTarget > 1, then content reconciliation may have to be done. For now, ignore subsequent responses.
+      }
+    }
+
+    /**
+     * This method is the entry point for handling responses received for requests sent out on behalf of this chunk.
+     * For gets, processing involves determining whether the request was successful, notifying the operation
+     * tracker so it can track the status of the operation, and notifying the response handler for failure detection.
+     * Finally, a check is done to determine whether the operation on the chunk is eligible for completion,
+     * if so the chunk operation is completed right away.
+     * @param responseInfo the response received for a request sent out on behalf of this chunk.
+     */
+    void handleResponse(ResponseInfo responseInfo) {
+      int correlationId = ((GetRequest) responseInfo.getRequest()).getCorrelationId();
+      // Get the GetOperation that generated the request.
+      GetRequestInfo getRequestInfo = correlationIdToGetRequestInfo.remove(correlationId);
+      if (getRequestInfo == null) {
+        // Ignore right away. This associated operation has completed.
```

Comment:
This is for `GetChunk` and it can absolutely happen. The situation is as follows:

1. `GetChunk` issues two requests with correlation ids 1 and 2.
2.  The first response comes in (say for request 1), and the chunk operation completes, and the chunk's state is cleared, the map itself is cleared.
3. The second response comes in, and it will no longer find correlation id 2 in the map, because the state of the `GetChunk is now different. In this case, the retrieved `GetRequestInfo` will be null.

Note that `GetBlobInfoOperation` does not have this problem as there are no chunks involved.